### PR TITLE
[python] dataloader optimization (picking up 1169)

### PR DESCRIPTION
**Pull request by ivirshup**
Building on top of #1169

This PR removes the new argument `method` (which was undocumented and had no tests atm) and instead using `return_sparse_X` to control conversion. Previously `return_sparse_X` controlled whether densification happened in a wrapping iterator, while now a dense array is created without having to make a CSR intermediate.

### CODE REVIEW

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:41]: nit (optional): why `npt.NDArray[Any]` and not `npt.NDArray[np.number]` (or something even more specific?)

Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:41]: Also, is the API signature really saying that it could return _any_ sparse type?  If it is actually constrained to a more specific set of sub-types (e.g, csr_matrix), it would be useful to further specify this type alias (even if just for the static type checking...)

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:113]: This creates a copy of the Arrow array which is wasteful/unnecessary.

Recommend instead doing a zero-copy conversion:
> row_indices_np = tbl.columns[0].to_numpy()

Same issue for all three columns.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:201]: IMHO, this motivates a dense option on the underlying blockwise iterator (i.e, in addition to the Table and scipy options). I don't think we should hold up this PR, but ideally just file a feature request GH issue in the TileDB-SOMA repo.

@ryan-williams - thoughts?

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:207]: nit: I would make this conditional on the same value used to select the iterator (e.g., `self.return_sparse_X`), and assert it is an ndarray.

Just for clarity - no other reason (i.e, the above code works AFAIK)

**Review by bkmartinjr — COMMENTED** [review]
Overall looks good - no major concerns. A few minor points of clarity and refinment noted inline.

In at least one case, there is a redundant data copy where zero-copy would suffice. Recommend tracing through the code to ensure no others (perhaps after this PR lands?), as reducing memory pressure is important in this use case.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py
Comment (ivirshup) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:207]: Yeah, I wasn't totally happy with the logic on this.

The type of the value should be invariant with the value of `return_sparse_X`. But that's not how it was in the original PR or the previous code. But I can go for consistency with other branches here.

Comment (ivirshup) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:205]: Its sorta weird that we make an iterator here but only ever use the first iteration

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py
Comment (ivirshup) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py]: 👍

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py
Comment (ivirshup) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py]: Hmm. This is opening up a big can of `typing` worms. There wouldn't be as much of an issue once we drop support for python 3.8.

I think I will go back to `Any` until we move to SPEC-0

**Review by ebezzi — APPROVED** [review]

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py
Comment (ryan-williams) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py]: Makes sense to me; we'd basically just move `_tables_to_np` into TileDB-SOMA?

**Pull request merged**
